nucypher/control/controllers.py

```diff
@@ -0,0 +1,341 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+import inspect
+import json
+from abc import ABC, abstractmethod
+from json import JSONDecodeError
+
+import maya
+from flask import Flask, Response
+from hendrix.deploy.base import HendrixDeploy
+from twisted.internet import reactor, stdio
+
+from nucypher.cli.processes import JSONRPCLineReceiver
+from nucypher.config.constants import MAX_UPLOAD_CONTENT_LENGTH
+from nucypher.control.emitters import StdoutEmitter, JSONRPCStdoutEmitter, WebEmitter
+from nucypher.control.interfaces import ControlInterface
+from nucypher.control.specifications.exceptions import SpecificationError
+from nucypher.exceptions import DevelopmentInstallationRequired
+from nucypher.network.resources import get_static_resources
+from nucypher.utilities.logging import Logger, GlobalLoggerSettings
+
+
+class ControllerBase(ABC):
+    """
+    A transactional interface for a human to interact with.
+    """
+    _emitter_class = NotImplemented
+
+    def __init__(self):
+        # Control Emitter
+        self.emitter = self._emitter_class()
+
+    def _perform_action(self, action: str, request: dict) -> dict:
+        """
+        This method is where input validation and method invocation
+        happens for all interface actions.
+        """
+        request = request or {}  # for requests with no input params request can be ''
```

Comment:
Should we add some sort of validation here to make sure that certain `action`s have the required `request`s?

Reply:
Great question. This is the beauty of the original "control" concept that I've used as a base. There is implicit validation that is performed on requests via a combination of the `marshmallow.fields` and `BaseSchema` concepts

The use of the `BaseSchema` concept defines the fields that make up the request, and the `marshmallow.fields` defines the validation for each field type. 

For the Schema definitions see:
- nucypher.utilities.porter.control.specifications.porter_schema.py (Porter)
- nucypher.characters.control.specifications.(alice.py, bob.py, enrico.py); for the original individual character schemas

For field validation see:
- nucypher.control.specifications.fields (for basic fields)
- nucypher.characters.control.specifications.fields (for fields needed for the character requests)
- nucypher.utilities.porter.control.specifications.fields (for fields needed  for Porter requests)

Reply:
Thanks for the explanation, that's a really cool take on schema validation :+1:

Reply:
It's pretty cool. The original "control" concept was created by @vepkenez and @KPrasch - I'm just refactoring it to make it non-character specific.